compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -27,26 +27,147 @@ pub enum AccessLevel {
 #[derive(Debug, Clone)]
 pub struct AccessLevels<Id = LocalDefId> {
-    pub map: FxHashMap<Id, AccessLevel>,
+    map: FxHashMap<Id, EffectiveVisibility>,
 }
 
-impl<Id: Hash + Eq> AccessLevels<Id> {
+impl<Id> AccessLevels<Id>
+where
+    Id: Hash + Eq + Copy,
+{
+    fn is_public_at_level(&self, id: Id, tag: AccessLevel) -> bool {
+        self.get_effective_vis(id).map_or(false, |effective_vis| effective_vis.is_public(tag))
+    }
+
     /// See `AccessLevel::Reachable`.
     pub fn is_reachable(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Reachable)
+        self.is_public_at_level(id, AccessLevel::Reachable)
     }
 
     /// See `AccessLevel::Exported`.
     pub fn is_exported(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Exported)
+        self.is_public_at_level(id, AccessLevel::Exported)
     }
 
     /// See `AccessLevel::Public`.
     pub fn is_public(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Public)
+        self.is_public_at_level(id, AccessLevel::Public)
+    }
+
+    pub fn is_some(&self, id: Id) -> bool {
+        self.is_public_at_level(id, AccessLevel::ReachableFromImplTrait)
+    }
+
+    pub fn get_access_level(&self, id: Id) -> Option<AccessLevel> {
+        self.get_effective_vis(id).and_then(|effective_vis| {
+            for level in [
+                AccessLevel::Public,
+                AccessLevel::Exported,
+                AccessLevel::Reachable,
+                AccessLevel::ReachableFromImplTrait,
+            ] {
+                if effective_vis.is_public(level) {
+                    return Some(level);
+                }
+            }
+            None
+        })
+    }
+
+    pub fn set_access_level(
+        &mut self,
+        id: Id,
+        access_level: Option<AccessLevel>,
+        tree: impl DefIdTree,
+    ) -> Option<AccessLevel> {
+        if let Some(tag) = access_level {
+            let mut effective_vis = self.get_effective_vis(id).copied().unwrap_or_default();
+            effective_vis.update(Visibility::Public, tag, tree);
+            self.set_effective_vis(id, effective_vis);
+        }
+        self.get_access_level(id)
+    }
+
+    pub fn set_effective_vis(&mut self, id: Id, effective_vis: EffectiveVisibility) {
+        self.map.insert(id, effective_vis);
+    }
+
+    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
+        self.map.get(&id)
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
+        self.map.iter()
+    }
+
+    fn map_id<OutId>(&self, f: impl Fn(Id) -> OutId) -> AccessLevels<OutId>
+    where
+        OutId: Hash + Eq + Copy,
+    {
+        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), *v)).collect() }
+    }
+}
+
+impl<Id> AccessLevels<Id>
+where
+    Id: Hash + Eq + Copy + Into<DefId>,
+{
+    pub fn to_def_id(&self) -> AccessLevels<DefId> {
```

Comment:
This function exists for other types like `LocalDefId` or `Visibility` because it's a common case.
for `AccessLevels` it should be enough to expose only `map_id` without additional convenience methods.